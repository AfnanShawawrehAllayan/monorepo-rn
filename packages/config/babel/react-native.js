module.exports = {
  presets: ['@react-native/babel-preset'],
  plugins: [
    [
      'module-resolver',
      {
        root: ['.'],
        extensions: [
          '.ios.ts',
          '.android.ts',
          '.ts',
          '.ios.tsx',
          '.android.tsx',
          '.tsx',
          '.jsx',
          '.js',
          '.json',
        ],
        alias: {
          '@components': './src/components',
          '@screens': './src/screens',
          '@navigation': './src/navigation',
          '@utils': './src/utils',
          '@hooks': './src/hooks',
          '@assets': './src/assets',
          '@config': './src/config',
        },
      },
    ],
    'react-native-reanimated/plugin',
  ],
  env: {
    production: {
      plugins: ['transform-remove-console'],
    },
  },
};
