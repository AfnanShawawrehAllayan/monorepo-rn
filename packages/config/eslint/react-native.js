module.exports = {
  root: true,
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:react/recommended',
    'plugin:react-hooks/recommended',
    'plugin:react-native/all',
    'plugin:import/errors',
    'plugin:import/warnings',
    'plugin:import/typescript',
    'plugin:jest/recommended',
    'prettier',
  ],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaFeatures: {
      jsx: true,
    },
    ecmaVersion: 2021,
    sourceType: 'module',
  },
  plugins: ['@typescript-eslint', 'react', 'react-native', 'import', 'jest'],
  settings: {
    react: {
      version: 'detect',
    },
    'import/resolver': {
      typescript: {},
      node: {
        extensions: ['.js', '.jsx', '.ts', '.tsx'],
      },
    },
  },
  rules: {
    // React Native specific rules
    'react-native/no-unused-styles': 'error',
    'react-native/split-platform-components': 'error',
    'react-native/no-inline-styles': 'warn',
    'react-native/no-color-literals': 'warn',
    'react-native/no-raw-text': ['warn', { skip: ['Text'] }],
    'react-native/no-single-element-style-arrays': 'error',

    // TypeScript rules
    '@typescript-eslint/explicit-module-boundary-types': 'warn',
    '@typescript-eslint/no-explicit-any': 'warn',
    '@typescript-eslint/no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
    '@typescript-eslint/consistent-type-imports': ['error', { prefer: 'type-imports' }],

    // React rules
    'react/prop-types': 'off',
    'react/react-in-jsx-scope': 'off',
    'react/display-name': 'off',
    'react/jsx-curly-brace-presence': ['error', { props: 'never', children: 'never' }],

    // Import rules
    'import/order': [
      'error',
      {
        groups: ['builtin', 'external', 'internal', 'parent', 'sibling', 'index'],
        'newlines-between': 'always',
        alphabetize: {
          order: 'asc',
          caseInsensitive: true,
        },
      },
    ],
    'import/no-duplicates': 'error',
    'import/no-unresolved': 'error',
    'import/no-cycle': 'error',

    // Jest rules
    'jest/no-disabled-tests': 'warn',
    'jest/no-focused-tests': 'error',
    'jest/no-identical-title': 'error',
    'jest/valid-expect': 'error',
  },
  env: {
    'react-native/react-native': true,
    jest: true,
    'jest/globals': true,
  },
  overrides: [
    {
      files: ['*.test.ts', '*.test.tsx', '*.spec.ts', '*.spec.tsx'],
      rules: {
        '@typescript-eslint/no-explicit-any': 'off',
        'react-native/no-raw-text': 'off',
      },
    },
    {
      files: ['*.js', '*.jsx'],
      rules: {
        '@typescript-eslint/explicit-module-boundary-types': 'off',
      },
    },
  ],
};
