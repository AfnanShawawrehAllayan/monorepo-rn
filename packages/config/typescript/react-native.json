{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "React Native",
  "compilerOptions": {
    "target": "esnext",
    "module": "esnext",
    "lib": ["esnext"],
    "allowJs": true,
    "jsx": "react-native",
    "noEmit": true,
    "isolatedModules": true,
    "strict": true,
    "moduleResolution": "node",
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "resolveJsonModule": true,
    "baseUrl": ".",
    "paths": {
      "*": ["*", "*.ios", "*.android"]
    },
    "forceConsistentCasingInFileNames": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "noUncheckedIndexedAccess": true,
    "noImplicitOverride": true,
    "noPropertyAccessFromIndexSignature": true,
    "exactOptionalPropertyTypes": true,
    "types": ["jest", "react-native"]
  },
  "exclude": ["node_modules", "babel.config.js", "metro.config.js", "jest.config.js"]
}
