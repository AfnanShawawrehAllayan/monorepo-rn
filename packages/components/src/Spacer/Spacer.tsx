import React from 'react';
import { View } from 'react-native';
import { useScale } from 'hooks';

type SpacerSize = 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'xxl' | number;
type SpacerDirection = 'horizontal' | 'vertical';

interface SpacerProps {
  size?: SpacerSize;
  direction?: SpacerDirection;
}

export const Spacer = ({ size = 'md', direction = 'vertical' }: SpacerProps): React.JSX.Element => {
  const { scale } = useScale();

  const getSize = (): number => {
    if (typeof size === 'number') return scale(size);
    
    switch (size) {
      case 'xs':
        return scale(4);
      case 'sm':
        return scale(8);
      case 'md':
        return scale(16);
      case 'lg':
        return scale(24);
      case 'xl':
        return scale(32);
      case 'xxl':
        return scale(48);
      default:
        return scale(16);
    }
  };

  const spacerSize = getSize();

  return (
    <View
      style={{
        width: direction === 'horizontal' ? spacerSize : undefined,
        height: direction === 'vertical' ? spacerSize : undefined,
      }}
    />
  );
}; 