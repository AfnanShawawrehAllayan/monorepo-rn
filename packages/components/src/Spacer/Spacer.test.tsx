// Mock modules before imports
jest.mock('react-native/Libraries/Utilities/Dimensions', () => ({
  get: jest.fn().mockReturnValue({ width: 375, height: 812 }),
}));

jest.mock('hooks', () => ({
  useScale: () => ({
    scale: jest.fn((value) => value),
    verticalScale: jest.fn((value) => value),
    moderateScale: jest.fn((value) => value),
  }),
}));

import React from 'react';
import { render } from '@testing-library/react-native';
import { Spacer } from './Spacer';

describe('Spacer Component', () => {
  it('renders correctly with default props', () => {
    const { toJSON } = render(<Spacer />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with custom numeric size', () => {
    const { toJSON } = render(<Spacer size={20} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with predefined sizes', () => {
    const sizes: ('xs' | 'sm' | 'md' | 'lg' | 'xl' | 'xxl')[] = ['xs', 'sm', 'md', 'lg', 'xl', 'xxl'];
    
    sizes.forEach(size => {
      const { toJSON } = render(<Spacer size={size} />);
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it('renders horizontal spacer', () => {
    const { toJSON } = render(<Spacer direction="horizontal" />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders vertical spacer', () => {
    const { toJSON } = render(<Spacer direction="vertical" />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders horizontal spacer with custom size', () => {
    const { toJSON } = render(<Spacer direction="horizontal" size={40} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders with all size variants and both directions', () => {
    const sizes: ('xs' | 'sm' | 'md' | 'lg' | 'xl' | 'xxl')[] = ['xs', 'sm', 'md', 'lg', 'xl', 'xxl'];
    const directions: ('horizontal' | 'vertical')[] = ['horizontal', 'vertical'];

    directions.forEach(direction => {
      sizes.forEach(size => {
        const { toJSON } = render(<Spacer direction={direction} size={size} />);
        expect(toJSON()).toMatchSnapshot();
      });
    });
  });
}); 