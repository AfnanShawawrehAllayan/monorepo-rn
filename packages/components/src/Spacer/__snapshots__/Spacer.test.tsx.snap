// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spacer Component renders correctly with default props 1`] = `
<View
  style={
    {
      "height": 16,
      "width": undefined,
    }
  }
/>
`;

exports[`Spacer Component renders horizontal spacer 1`] = `
<View
  style={
    {
      "height": undefined,
      "width": 16,
    }
  }
/>
`;

exports[`Spacer Component renders horizontal spacer with custom size 1`] = `
<View
  style={
    {
      "height": undefined,
      "width": 40,
    }
  }
/>
`;

exports[`Spacer Component renders vertical spacer 1`] = `
<View
  style={
    {
      "height": 16,
      "width": undefined,
    }
  }
/>
`;

exports[`Spacer Component renders with all size variants and both directions 1`] = `
<View
  style={
    {
      "height": undefined,
      "width": 4,
    }
  }
/>
`;

exports[`Spacer Component renders with all size variants and both directions 2`] = `
<View
  style={
    {
      "height": undefined,
      "width": 8,
    }
  }
/>
`;

exports[`Spacer Component renders with all size variants and both directions 3`] = `
<View
  style={
    {
      "height": undefined,
      "width": 16,
    }
  }
/>
`;

exports[`Spacer Component renders with all size variants and both directions 4`] = `
<View
  style={
    {
      "height": undefined,
      "width": 24,
    }
  }
/>
`;

exports[`Spacer Component renders with all size variants and both directions 5`] = `
<View
  style={
    {
      "height": undefined,
      "width": 32,
    }
  }
/>
`;

exports[`Spacer Component renders with all size variants and both directions 6`] = `
<View
  style={
    {
      "height": undefined,
      "width": 48,
    }
  }
/>
`;

exports[`Spacer Component renders with all size variants and both directions 7`] = `
<View
  style={
    {
      "height": 4,
      "width": undefined,
    }
  }
/>
`;

exports[`Spacer Component renders with all size variants and both directions 8`] = `
<View
  style={
    {
      "height": 8,
      "width": undefined,
    }
  }
/>
`;

exports[`Spacer Component renders with all size variants and both directions 9`] = `
<View
  style={
    {
      "height": 16,
      "width": undefined,
    }
  }
/>
`;

exports[`Spacer Component renders with all size variants and both directions 10`] = `
<View
  style={
    {
      "height": 24,
      "width": undefined,
    }
  }
/>
`;

exports[`Spacer Component renders with all size variants and both directions 11`] = `
<View
  style={
    {
      "height": 32,
      "width": undefined,
    }
  }
/>
`;

exports[`Spacer Component renders with all size variants and both directions 12`] = `
<View
  style={
    {
      "height": 48,
      "width": undefined,
    }
  }
/>
`;

exports[`Spacer Component renders with custom numeric size 1`] = `
<View
  style={
    {
      "height": 20,
      "width": undefined,
    }
  }
/>
`;

exports[`Spacer Component renders with predefined sizes 1`] = `
<View
  style={
    {
      "height": 4,
      "width": undefined,
    }
  }
/>
`;

exports[`Spacer Component renders with predefined sizes 2`] = `
<View
  style={
    {
      "height": 8,
      "width": undefined,
    }
  }
/>
`;

exports[`Spacer Component renders with predefined sizes 3`] = `
<View
  style={
    {
      "height": 16,
      "width": undefined,
    }
  }
/>
`;

exports[`Spacer Component renders with predefined sizes 4`] = `
<View
  style={
    {
      "height": 24,
      "width": undefined,
    }
  }
/>
`;

exports[`Spacer Component renders with predefined sizes 5`] = `
<View
  style={
    {
      "height": 32,
      "width": undefined,
    }
  }
/>
`;

exports[`Spacer Component renders with predefined sizes 6`] = `
<View
  style={
    {
      "height": 48,
      "width": undefined,
    }
  }
/>
`;
